#DATA WRANGLING

*  It is the process of cleaning, transforming, and structuring raw data into a usable format for analysis.






**Step1** - Collecting the data and loading data

In [2]:
import pandas as pd
df = pd.read_csv("RewardsData.csv")  # Loading CSV file

#Printing the first 6 records
df.head(6)

User ID  Birthdate           City     State    Zip  Available Points  \
0        1  1/31/1992   Minneapolis    Georgia  55401               300   
1        2        NaN   Los Angeles   Colorado  90210               320   
2        3  1/11/1990         Merida        AL  97204               330   
3        4  4/16/1990  Winston-Salem        NC  27106             10790   
4        5  3/24/1994  Winston-Salem        NC  27109               650   
5        6  10/9/1995              G        AL  85588              7085   

   Total Points Earned  Points Spent         Joined On         Last Seen  \
0                  300             0     9/4/2020 0:15  12/16/2021 17:42   
1                  320             0   8/19/2020 11:15   4/13/2022 14:28   
2                  330             0  10/13/2020 16:20   4/21/2021 10:54   
3                14290          3500   8/18/2020 13:21   6/30/2022 22:22   
4                  650             0   8/21/2020 11:27  12/30/2020 10:57   
5                 7395           310   8/18/2020 14:28    6/8/2022 23:55   

                                     Tags  
0                             ClassOf2025  
1                       ClassOf2025|apple  
2                      Freshman|Developer  
3         GraduateStudent|admin|MBBvsDuke  
4  Student|NOBSBvsNDGame3|NOBSBvsNDGame3A  
5                   GraduateStudent|admin

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7252 entries, 0 to 7251
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   User ID              7252 non-null   int64 
 1   Birthdate            1172 non-null   object
 2   City                 330 non-null    object
 3   State                330 non-null    object
 4   Zip                  330 non-null    object
 5   Available Points     7252 non-null   int64 
 6   Total Points Earned  7252 non-null   int64 
 7   Points Spent         7252 non-null   int64 
 8   Joined On            7252 non-null   object
 9   Last Seen            7031 non-null   object
 10  Tags                 7252 non-null   object
dtypes: int64(4), object(7)
memory usage: 623.3+ KB


### Finding Missing Values

We have 2 functions in pandas isnull() and isna() to identify missing

*   We have 2 functions in pandas isnull() and isna() to identify missing values in DataFrame.
*   It returns True for missing values, False for non-missing values.

In [34]:
df.isnull()

User ID  Birthdate   City  State    Zip  Available Points  \
0       False      False  False  False  False             False   
1       False       True  False  False  False             False   
2       False      False  False  False  False             False   
3       False      False  False  False  False             False   
4       False      False  False  False  False             False   
...       ...        ...    ...    ...    ...               ...   
7247    False       True   True   True   True             False   
7248    False       True   True   True   True             False   
7249    False      False   True   True   True             False   
7250    False       True   True   True   True             False   
7251    False       True   True   True   True             False   

      Total Points Earned  Points Spent  Joined On  Last Seen   Tags  
0                   False         False      False      False  False  
1                   False         False      False      False  False  
2                   False         False      False      False  False  
3                   False         False      False      False  False  
4                   False         False      False      False  False  
...                   ...           ...        ...        ...    ...  
7247                False         False      False      False  False  
7248                False         False      False      False  False  
7249                False         False      False      False  False  
7250                False         False      False      False  False  
7251                False         False      False      False  False  

[7252 rows x 11 columns]

In [24]:
df.isna().sum() #Counting missing values per column

User ID                   0
Birthdate              6080
City                   6922
State                  6922
Zip                    6922
Available Points          0
Total Points Earned       0
Points Spent              0
Joined On                 0
Last Seen               221
Tags                      0
dtype: int64

In [25]:
print(len(df))  # Number of rows in the dataset

7252


###Handling missing values by dropping them.

We have dropna() function in pandas where it will drop the entire row if it contains at least one NaN (missing) value.

In [26]:
df.dropna(inplace=True)

In [27]:
df.isna().sum()

User ID                0
Birthdate              0
City                   0
State                  0
Zip                    0
Available Points       0
Total Points Earned    0
Points Spent           0
Joined On              0
Last Seen              0
Tags                   0
dtype: int64

In [29]:
print(len(df))  # Number of rows after droping null and NAN value rows

172


#### So we can see that there is a huge loss of data by dropping the NAN value columns using dropna() function. So we have different methods to deal with missing values come lets have a look of them.

### Handing missing values by imputing


I have reloaded the data set so we have complete data instead of 172 rows.

In [68]:
df.isnull().sum()

User ID                   0
Birthdate              6080
City                   6922
State                  6922
Zip                    6922
Available Points          0
Total Points Earned       0
Points Spent              0
Joined On                 0
Last Seen               221
Tags                      0
dtype: int64

We can see clearly that the columns named "Birthdate", "City", "State", "Zip", "Last Seen" has missing values.


In these columns "Birthdate" is a date column, whereas "LastSeen" is a datetime format column

First lets see how to do imputation for these kind of columns.

In [69]:
df['Birthdate'].fillna(df['Birthdate'].mode()[0], inplace=True)  # Most common birthdate
df.head(5)

<ipython-input-69-6d4a8b09a731>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Birthdate'].fillna(df['Birthdate'].mode()[0], inplace=True)  # Most common birthdate


User ID   Birthdate           City     State    Zip  Available Points  \
0        1   1/31/1992   Minneapolis    Georgia  55401               300   
1        2  12/15/1999   Los Angeles   Colorado  90210               320   
2        3   1/11/1990         Merida        AL  97204               330   
3        4   4/16/1990  Winston-Salem        NC  27106             10790   
4        5   3/24/1994  Winston-Salem        NC  27109               650   

   Total Points Earned  Points Spent         Joined On         Last Seen  \
0                  300             0     9/4/2020 0:15  12/16/2021 17:42   
1                  320             0   8/19/2020 11:15   4/13/2022 14:28   
2                  330             0  10/13/2020 16:20   4/21/2021 10:54   
3                14290          3500   8/18/2020 13:21   6/30/2022 22:22   
4                  650             0   8/21/2020 11:27  12/30/2020 10:57   

                                     Tags  
0                             ClassOf2025  
1                       ClassOf2025|apple  
2                      Freshman|Developer  
3         GraduateStudent|admin|MBBvsDuke  
4  Student|NOBSBvsNDGame3|NOBSBvsNDGame3A

We can clearly see that the 2nd row Birthdate was NAN value before now we have Imputed it with Most common birthdate in the dataset.

In [70]:
df['Last Seen'] = pd.to_datetime(df['Last Seen'], errors='coerce')  # Convert to datetime
median_last_seen = df['Last Seen'].median()  # Compute the median datetime
df['Last Seen'] = df['Last Seen'].fillna(median_last_seen)  # Replace NaT with the median date

Above code cell replaces missing "Last Seen" values with the middle timestamp.

In [71]:
df.isnull().sum()

User ID                   0
Birthdate                 0
City                   6922
State                  6922
Zip                    6922
Available Points          0
Total Points Earned       0
Points Spent              0
Joined On                 0
Last Seen                 0
Tags                      0
dtype: int64

We have 3 more columns with NAN values.

We have "City", "State" categorical columns.

In [72]:
# For categorical columns, fill missing values with the most frequent value (mode)
df['City'] = df['City'].fillna(df['City'].mode()[0])
df['State'] = df['State'].fillna(df['State'].mode()[0])

In [73]:
df.isnull().sum()

User ID                   0
Birthdate                 0
City                      0
State                     0
Zip                    6922
Available Points          0
Total Points Earned       0
Points Spent              0
Joined On                 0
Last Seen                 0
Tags                      0
dtype: int64

We have 1 more column "Zip" we can use K-Nearest Neighbors (KNN) imputation to predict missing zip codes based on similar rows (similarity can be determined by other features like city, state, or available points). The KNNImputer from sklearn can be used for this



In [74]:
from sklearn.impute import KNNImputer

# Convert 'Zip' to numerical format for imputation (if necessary)
df['Zip'] = pd.to_numeric(df['Zip'], errors='coerce')

# Apply KNN imputation
imputer = KNNImputer(n_neighbors=5)  # You can change the number of neighbors
df['Zip'] = imputer.fit_transform(df[['Zip']])


In [75]:
df.isnull().sum()

User ID                0
Birthdate              0
City                   0
State                  0
Zip                    0
Available Points       0
Total Points Earned    0
Points Spent           0
Joined On              0
Last Seen              0
Tags                   0
dtype: int64

**Hooray! We have successfully handled all null values! 🎉**

There are so many techniques techniques to handle missing values we can perform all those based on the use case.

####Now lets find out wheather our dataset has duplicates or not.
####Duplicate data can cause bias in analysis.




In [3]:
df.duplicated().sum()  # Count duplicate rows

10

We have detected there are 10 rows with duplicate values lets remove the duplicated rows.

In [4]:
df.drop_duplicates(inplace=True)  # Remove duplicate rows

In [5]:
df.duplicated().sum()  # Count duplicate rows

0

We have successfully removed the duplicated rows from our data.

###Now lets fix the datatypes of our data

In [6]:
df.dtypes  # Check column data types

User ID                 int64
Birthdate              object
City                   object
State                  object
Zip                    object
Available Points        int64
Total Points Earned     int64
Points Spent            int64
Joined On              object
Last Seen              object
Tags                   object
dtype: object

In [8]:
df['Last Seen'] = pd.to_datetime(df['Last Seen'])

In [9]:
df.dtypes

User ID                         int64
Birthdate                      object
City                           object
State                          object
Zip                            object
Available Points                int64
Total Points Earned             int64
Points Spent                    int64
Joined On                      object
Last Seen              datetime64[ns]
Tags                           object
dtype: object